File: examples/java/SaleListing/src/main/java/com/umatrangolo/giltapi/example/SaleListingApp.java
package com.umatrangolo.giltapi.example;

import com.umatrangolo.giltapi.Products;
import com.umatrangolo.giltapi.Sales;

import com.umatrangolo.giltapi.model.Sale;
import com.umatrangolo.giltapi.model.Product;
import com.umatrangolo.giltapi.model.Store;

import scala.concurrent.Await;

import static com.umatrangolo.giltapi.client.GiltClientFactory.*;
import scala.concurrent.duration.Duration;

import java.util.List;

public class SaleListingApp {
    public static void main(String[] args) throws Exception {
        System.out.println("**** Fetching sale listing ... ");

        Sales salesClient = newSalesClientInstance();
        Products productClient = newProductsClientInstance();

        // active sales
        System.out.println("======== ACTIVE SALES ========");
        List<Sale> activeSales = salesClient.getActiveSales().get();
        for (Sale sale: activeSales) {
            System.out.println(">> " + sale.getName());
        }

        System.out.println("======== UPCOMING SALES ========");
        List<Sale> upcomingSales = salesClient.getUpcomingSales().get();
        for (Sale sale: upcomingSales) {
            System.out.println(">> " + sale.getName());
        }

        // details about a single sale in the Women store
        System.out.println("======== FIRST SALE FOR WOMEN ========");
        List<Sale> salesForWomen = salesClient.getActiveSales(Store.Women).get();
        if (!salesForWomen.isEmpty()) {
            Sale sale = salesForWomen.get(0);
            System.out.println(">> Getting details about " + sale.getName());
            Sale details = salesClient.getSale(sale.getKey(), Store.Women).get().get();
            System.out.println(">> [%s] is \n%s".format(sale.getKey(), details));
        }

        // browsing a product
        System.out.println("======== BROWSING PRODUCT ========");
        if (!salesForWomen.isEmpty()) {
            Sale sale = salesForWomen.get(0);
            List<Long> productIds = sale.getProductIds();
            if (!productIds.isEmpty()) {
                Product product = productClient.getProduct(productIds.get(0)).get().get();
                System.out.println("---- Product with id " + product);
            }
        }

        System.out.println("======== BYE ========");
    }
}

File: examples/scala/SaleListing/src/main/scala/com/umatrangolo/giltapi/example/SaleListingApp.scala
package com.umatrangolo.giltapi.example

import com.google.common.base.Optional

import com.umatrangolo.giltapi.client.GiltClientFactory
import com.umatrangolo.giltapi.model._
import com.umatrangolo.giltapi.{ Sales, Products }
import com.umatrangolo.giltapi.utils.FutureConversions._
import com.umatrangolo.giltapi.utils.GuavaConversions._

import java.util.{ List => JList }

import scala.collection.JavaConversions._
import scala.concurrent._
import scala.concurrent.duration._
import scala.language.postfixOps

import ExecutionContext.Implicits.global

object SaleListingApp extends App {
  val salesClient = GiltClientFactory.newSalesClientInstance()
  val productClient = GiltClientFactory.newProductsClientInstance()

  salesClient.getActiveSales(Store.Women).map { salesForWomen =>

    // details about a single sale in the Women store
    println("======== FIRST SALE FOR WOMEN ========")
    salesForWomen.headOption.foreach { sale =>
      println(">> Getting details about " + sale.name)
      salesClient.getSale(sale.key, Store.Women).map { saleDetails =>
        println(">> [%s] is \n%s".format(sale.key, saleDetails))
      }
    }

    // browsing a product
    println("======== BROWSING PRODUCT ========")
    salesForWomen.headOption.foreach { sale =>
      println(">> first sale in the Women store is " + sale.name)
      sale.productIds.headOption.foreach { pid =>
        println("\n\n**** Getting info about Product with id: " + pid)
        productClient.getProduct(pid).map { product =>
          product.foreach { p => println("---- Product with id %s is: \n%s".format(pid, p)) }
          println("======== BYE ========")
        }
      }
    }
  }
}
